public/app/features/dashboard/state/PanelModel.ts

```diff
@@ -280,16 +281,21 @@ export class PanelModel {
     this.cachedPluginOptions[oldPluginId] = oldOptions;
     this.restorePanelOptions(pluginId);
 
-    // switch
-    this.type = pluginId;
-    this.plugin = newPlugin;
-    this.applyPluginOptionDefaults(newPlugin);
     // Let panel plugins inspect options from previous panel and keep any that it can use
     if (newPlugin.onPanelTypeChanged) {
+      let old: any = {};
+      if (wasAngular) {
+        old = { angular: oldOptions };
+      } else if (oldOptions && oldOptions.options) {
+        old = oldOptions.options;
+      }
       this.options = this.options || {};
-      const old = oldOptions && oldOptions.options ? oldOptions.options : {};
       Object.assign(this.options, newPlugin.onPanelTypeChanged(this.options, oldPluginId, old));
     }
+    // switch
+    this.type = pluginId;
+    this.plugin = newPlugin;
+    this.applyPluginOptionDefaults(newPlugin);
```

Comment:
Biggest change here is that `applyPluginOptionDefaults` is called **after** newPlugin.onPanelTypeChanged

This also passes the whole panel object in when switching from angular panels (the angular options are not really stored on `oldOptions.options`, rather in `oldOptions` directly)